### ISSUE TITLE
[Feature] Integrate plotly for graphs

### ISSUE THREAD

**Issue by mohith01** [gradio-app/gradio#141]
Gradio is great but it would be really amazing if we can add graphs using plotly or matplotlib in there.

**Comment by mohith01** [comment]
If anyone comes across this problem, we can solve this by exporting plotly picture as html and then importing it as html in gradio

**Comment by mohith01** [comment]
```py
#delimiter is space
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import codecs
from bs4 import BeautifulSoup

def fun2(a,b):
  fig = go.Figure()
  fig.add_trace(go.Scatter(x=list(map(float,a.split(' '))), y=list(map(float,b.split(' ' ))),mode='lines+markers',name='hi',marker=dict(size=10,line=dict(width=2))))

  fig.write_html("test1.html")
  f = codecs.open("test1.html",'r','utf-8')
  doc = BeautifulSoup(f)
  return str(doc)


gr.Interface(fun2,[gr.inputs.Textbox(default='1 2 3 4'),gr.inputs.Textbox(default='4 2 1 3')],gr.outputs.HTML()).launch()
```

Your output will be something like this
![](https://user-images.githubusercontent.com/42903811/114353298-80956a80-9b8a-11eb-8fa8-a33b3c5c98f1.png)

**Comment by discdiver** [comment]
This hack for plotly does not work for me in gradio 2.2.15  with python 3.9.  The image is saved to file, but won't display on read in - fails silently without an error in the console.

**Comment by mohith01** [comment]
> This hack for plotly does not work for me in gradio 2.2.15 with python 3.9. The image is saved to file, but won't display on read in - fails silently without an error in the console.

So I checked this with python 3.8 too and it is giving me the same as what you have said ryt now. However when I inspected the console, It gives me a "DOM exception option was aborted" error and there;s some problem with analytical logging. I think there's some word limit in HTML of Gradio.. I think we can do the following.. so plotly [export](https://plotly.com/python/interactive-html-export/) has a lot of options such as plolty js with cdn and such.. Ig we can experiment with that and see if we can reduce the file size.. Also I think we should tag the moderators for integrating plotly

**Comment by P8H** [comment]
I noticed that any javascript that is injected by the output HTML component is not executed.
Using plotly in the HTML version requires JS to be executed, also because the data for the plot is encoded in the JS script element.

A word limit seems to be here not the problem, at least in my case. I was able to find the complete code, coming from `fig.write_html("test1.html", include_plotlyjs='cdn')` in the DOM.

I assume, that the HTML output is added into the DOM with a JS method like `.innerHTML()` which prevents executing of script elements. Here is a SO post about this kind of problem: https://security.stackexchange.com/a/240362

It would be amazing to see plotly plots in Gradio Apps.

**Comment by pandya6988** [comment]
```
def get_plot(model_name, chunk, version='training', label='senti'):
    try:
        with open(f'data/{model_name}', 'rb') as f:
            model_dict = pickle.load(f)
    except: 
        error_msg = bc.bcolors.FAIL + 'Unvalid file.'
        return error_msg 
    
    pca2 = PCA(n_components=2)
    pca2_chunk = pca2.fit_transform(model_dict[version]['encoded_text'][chunk])
    df = pd.DataFrame()
    df['x'] = pca2_chunk[:,0].squeeze().tolist()
    df['y'] = pca2_chunk[:,1].squeeze().tolist()
    df['language'] = model_dict[version]['language']
    df['senti'] = model_dict[version]['sentiment']
    df['text'] = [text[:250]for text in model_dict[version]['text']]

    fig = plx.scatter(df, x='x', y='y', color=label, title=f'PCA {model_name} {version} {chunk}', hover_data=['text'], opacity=0.2)
    
    print(bc.bcolors.HEADER + "Converting to html ..." + bc.bcolors.ENDC)
    fig.write_html("test1.html", include_plotlyjs='cdn')
    f = codecs.open("test1.html",'r','utf-8')
    doc = BeautifulSoup(f)
    print(bc.bcolors.HEADER + "Done..." + bc.bcolors.ENDC)
    return str(doc)

interface = gr.Interface(get_plot, inputs=[inputs, chunk, version, labels], outputs=gr.outputs.HTML())
interface.launch(debug=True)
```

This is not working. anyone can please help!!!

**Comment by abidlabs** [comment]
An integration with Observable to support https://github.com/observablehq/plot in output components would be nice as well.

**Comment by abidlabs** [comment]
Plotly is now supported with the new `Plot` component as of version 2.9.3!

Closed via #805
